[CI] EsExecutorsTests testFixedBoundedRejectOnShutdown failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/5d6uea2bizhik/tests/:server:test/org.elasticsearch.common.util.concurrent.EsExecutorsTests/testFixedBoundedRejectOnShutdown


**Reproduction line:**
```
./gradlew ':server:test' --tests "org.elasticsearch.common.util.concurrent.EsExecutorsTests.testFixedBoundedRejectOnShutdown" -Dtests.seed=988CB26CB6520078 -Dtests.locale=es-CR -Dtests.timezone=CET -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
[Failure dashboard for `org.elasticsearch.common.util.concurrent.EsExecutorsTests#testFixedBoundedRejectOnShutdown`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testFixedBoundedRejectOnShutdown'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.common.util.concurrent.EsExecutorsTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))


**Failure excerpt:**
```
java.lang.AssertionError: (No message provided)

  at org.junit.Assert.fail(Assert.java:87)
  at org.junit.Assert.assertTrue(Assert.java:42)
  at org.junit.Assert.assertTrue(Assert.java:53)
  at org.elasticsearch.common.util.concurrent.EsExecutorsTests.assertShutdownAndRejectingTasks(EsExecutorsTests.java:669)
  at org.elasticsearch.common.util.concurrent.EsExecutorsTests.runRejectOnShutdownTest(EsExecutorsTests.java:636)
  at org.elasticsearch.common.util.concurrent.EsExecutorsTests.testFixedBoundedRejectOnShutdown(EsExecutorsTests.java:604)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

--- a/server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java
+++ b/server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java
@@ -23,7 +23,7 @@ public class EsAbortPolicy extends EsRejectedExecutionHandler {
                         Thread.currentThread().interrupt();
                         throw new IllegalStateException("forced execution, but got interrupted", e);
                     }
-                    if (executor.isShutdown() == false || sizeBlockingQueue.remove(r) == false) {
+                    if ((executor.isShutdown() && sizeBlockingQueue.remove(r)) == false) {
                         return;
                     } // else fall through and reject the task since the executor is shut down
                 } else {

--- a/server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java
+++ b/server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java
@@ -23,11 +23,9 @@ public class EsAbortPolicy extends EsRejectedExecutionHandler {
                         Thread.currentThread().interrupt();
                         throw new IllegalStateException("forced execution, but got interrupted", e);
                     }
-                    // we need to check again the executor state as it might have been concurrently shut down; in this case
-                    // the executor's workers are shutting down and might have already picked up the task for execution.
                     if (executor.isShutdown() == false || sizeBlockingQueue.remove(r) == false) {
                         return;
-                    }
+                    } // else fall through and reject the task since the executor is shut down
                 } else {
                     throw new IllegalStateException("expected but did not find SizeBlockingQueue: " + executor);
                 }

--- a/server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java
+++ b/server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java
@@ -13,20 +13,27 @@ import java.util.concurrent.ThreadPoolExecutor;
 public class EsAbortPolicy extends EsRejectedExecutionHandler {
 
     @Override
-    public void rejectedExecution(Runnable task, ThreadPoolExecutor executor) {
-        if (executor.isShutdown() == false && task instanceof AbstractRunnable r && r.isForceExecution()) {
-            if (executor.getQueue() instanceof SizeBlockingQueue<Runnable> sizeBlockingQueue) {
-                try {
-                    sizeBlockingQueue.forcePut(task);
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    throw new IllegalStateException("forced execution, but got interrupted", e);
+    public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+        if (executor.isShutdown() == false && r instanceof AbstractRunnable abstractRunnable) {
+            if (abstractRunnable.isForceExecution()) {
+                if (executor.getQueue() instanceof SizeBlockingQueue<Runnable> sizeBlockingQueue) {
+                    try {
+                        sizeBlockingQueue.forcePut(r);
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new IllegalStateException("forced execution, but got interrupted", e);
+                    }
+                    // we need to check again the executor state as it might have been concurrently shut down; in this case
+                    // the executor's workers are shutting down and might have already picked up the task for execution.
+                    if (executor.isShutdown() == false || sizeBlockingQueue.remove(r) == false) {
+                        return;
+                    }
+                } else {
+                    throw new IllegalStateException("expected but did not find SizeBlockingQueue: " + executor);
                 }
-            } else {
-                throw new IllegalStateException("expected but did not find SizeBlockingQueue: " + executor);
             }
         }
         incrementRejections();
-        throw newRejectedException(task, executor, executor.isShutdown());
+        throw newRejectedException(r, executor, executor.isShutdown());
     }
 }

--- a/server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java
+++ b/server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java
@@ -13,25 +13,20 @@ import java.util.concurrent.ThreadPoolExecutor;
 public class EsAbortPolicy extends EsRejectedExecutionHandler {
 
     @Override
-    public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
-        if (r instanceof AbstractRunnable abstractRunnable) {
-            if (abstractRunnable.isForceExecution()) {
-                if (executor.getQueue() instanceof SizeBlockingQueue<Runnable> sizeBlockingQueue) {
-                    try {
-                        sizeBlockingQueue.forcePut(r);
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new IllegalStateException("forced execution, but got interrupted", e);
-                    }
-                    if ((executor.isShutdown() && sizeBlockingQueue.remove(r)) == false) {
-                        return;
-                    } // else fall through and reject the task since the executor is shut down
-                } else {
-                    throw new IllegalStateException("expected but did not find SizeBlockingQueue: " + executor);
+    public void rejectedExecution(Runnable task, ThreadPoolExecutor executor) {
+        if (executor.isShutdown() == false && task instanceof AbstractRunnable r && r.isForceExecution()) {
+            if (executor.getQueue() instanceof SizeBlockingQueue<Runnable> sizeBlockingQueue) {
+                try {
+                    sizeBlockingQueue.forcePut(task);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new IllegalStateException("forced execution, but got interrupted", e);
                 }
+            } else {
+                throw new IllegalStateException("expected but did not find SizeBlockingQueue: " + executor);
             }
         }
         incrementRejections();
-        throw newRejectedException(r, executor, executor.isShutdown());
+        throw newRejectedException(task, executor, executor.isShutdown());
     }
 }

